zcash_client_sqlite/src/wallet/common.rs

```diff
@@ -248,33 +258,181 @@ pub(crate) fn count_outputs(
+        account: AccountId,
+        selector: &NoteFilter,
+        chain_tip_height: BlockHeight,
+    ) -> Result<Option<NonNegativeAmount>, SqliteClientError> {
+        match selector {
+            NoteFilter::ExceedsMinValue(v) => Ok(Some(*v)),
+            NoteFilter::ExceedsPriorSendPercentile(n) => {
+                let mut bucket_query = conn.prepare(
+                    "WITH bucketed AS (
+                        SELECT s.value, NTILE(10) OVER (ORDER BY s.value) AS bucket_index
+                        FROM sent_notes s
+                        JOIN transactions t ON s.tx = t.id_tx
+                        WHERE s.from_account_id = :account_id
+                        -- only count mined transactions
+                        AND t.mined_height IS NOT NULL
+                        -- exclude change and account-internal sends
+                        AND (s.to_account_id IS NULL OR s.from_account_id != s.to_account_id)
+                    )
+                    SELECT MAX(value) as value
+                    FROM bucketed
+                    GROUP BY bucket_index
+                    ORDER BY bucket_index",
+                )?;
+
+                let bucket_maxima = bucket_query
+                    .query_and_then::<_, SqliteClientError, _, _>(
+                        named_params![":account_id": account.0],
+                        |row| {
+                            NonNegativeAmount::from_nonnegative_i64(row.get::<_, i64>(0)?).map_err(
+                                |_| {
+                                    SqliteClientError::CorruptedData(format!(
+                                        "Negative received note value: {}",
+                                        n.value()
+                                    ))
+                                },
+                            )
+                        },
+                    )?
+                    .collect::<Result<Vec<_>, _>>()?;
+
+                // Pick a bucket index by scaling the requested percentile to the number of buckets
+                let i = (bucket_maxima.len() * usize::from(*n) / 100).saturating_sub(1);
+                Ok(bucket_maxima.get(i).copied())
+            }
+            NoteFilter::ExceedsBalancePercentage(p) => {
+                let balance = conn.query_row_and_then::<_, SqliteClientError, _, _>(
+                    &format!(
+                        "SELECT SUM(rn.value)
+                         FROM v_received_outputs rn
+                         INNER JOIN accounts a ON a.id = rn.account_id
+                         INNER JOIN transactions ON transactions.id_tx = rn.transaction_id
+                         WHERE rn.account_id = :account_id
+                         AND a.ufvk IS NOT NULL
+                         AND transactions.mined_height IS NOT NULL
+                         AND rn.pool != :transparent_pool
+                         AND rn.id_within_pool_table NOT IN (
+                           SELECT rns.received_output_id
```

Comment:
Does SQLite generate the transient sub-table on the fly for every row, or can it share the sub-table across rows? If the latter, then instead of having the sub-table be row-specific with `WHERE rns.pool == rn.pool`, could we get better performance with something like:
```suggestion
                         AND (rn.pool, rn.id_within_pool_table) NOT IN (
                           SELECT rns.pool, rns.received_output_id
```

Reply:
Good question, I haven't looked at the query plan.

Reply:
Tested and this works.